cmd/kanikoExecute.go

```diff
@@ -134,117 +135,207 @@ func runKanikoExecute(config *kanikoExecuteOptions, telemetryData *telemetry.Cus
+					return fmt.Errorf("failed to build image '%v' using '%v': %w", containerImageName, config.DockerfilePath, err)
+				}
+
+				commonPipelineEnvironment.container.imageNameTags = append(commonPipelineEnvironment.container.imageNameTags, containerImageNameTag)
+				commonPipelineEnvironment.container.imageNames = append(commonPipelineEnvironment.container.imageNames, containerImageName)
 			}
-			// errors are already caught with previous call to docker.ContainerRegistryFromImage
-			containerImageName, _ := docker.ContainerImageNameFromImage(config.ContainerImage)
-			containerImageNameTag, _ := docker.ContainerImageNameTagFromImage(config.ContainerImage)
-			dest = []string{"--destination", config.ContainerImage}
-			commonPipelineEnvironment.container.registryURL = fmt.Sprintf("https://%v", containerRegistry)
-			commonPipelineEnvironment.container.imageNameTag = containerImageNameTag
-			commonPipelineEnvironment.container.imageNameTags = append(commonPipelineEnvironment.container.imageNameTags, containerImageNameTag)
-			commonPipelineEnvironment.container.imageNames = append(commonPipelineEnvironment.container.imageNames, containerImageName)
 		}
-		config.BuildOptions = append(config.BuildOptions, dest...)
-	} else {
-		log.Entry().Infof("Running Kaniko build with destination defined via buildOptions: %v", config.BuildOptions)
 
-		destination := ""
+		// for compatibility reasons also fill single imageNameTag field with "root" image in commonPipelineEnvironment
+		containerImageNameAndTag := fmt.Sprintf("%v:%v", config.ContainerImageName, config.ContainerImageTag)
+		commonPipelineEnvironment.container.imageNameTag = containerImageNameAndTag
+		commonPipelineEnvironment.container.registryURL = config.ContainerRegistryURL
+
+		if config.CreateBOM {
+			// Syft for multi image, generates bom-docker-(1/2/3).xml
+			return syft.GenerateSBOM(config.SyftDownloadURL, "/kaniko/.docker", execRunner, fileUtils, httpClient, commonPipelineEnvironment.container.registryURL, commonPipelineEnvironment.container.imageNameTags)
+		}
+		return nil
+
+	case piperutils.ContainsString(config.BuildOptions, "--destination"):
+		log.Entry().Infof("Running Kaniko build with destination defined via buildOptions: %v", config.BuildOptions)
 
 		for i, o := range config.BuildOptions {
 			if o == "--destination" && i+1 < len(config.BuildOptions) {
-				destination = config.BuildOptions[i+1]
-				break
+				destination := config.BuildOptions[i+1]
+
+				containerRegistry, err := docker.ContainerRegistryFromImage(destination)
+				if err != nil {
+					log.SetErrorCategory(log.ErrorConfiguration)
+					return errors.Wrapf(err, "invalid registry part in image %v", destination)
+				}
+				if commonPipelineEnvironment.container.registryURL == "" {
+					commonPipelineEnvironment.container.registryURL = fmt.Sprintf("https://%v", containerRegistry)
+				}
```

Comment:
What is the reason to have if statement here? Do we expect that the `commonPipelineEnvironment.container.registryURL` can be filled somewhere else?

Reply:
we iterate `BuildOptions` `--destination` elements, and because commonPipelineEnvironment.container.registryURL is a single string, it is meaningless to overwrite it on each iteration.